// ipfs/lib/cli/commands/files/get.js
'use strict';

var debug = require('debug');
var utils = require('../../utils');
var log = debug('cli:files');
log.error = debug('cli:files:error');
var fs = require('fs');
var path = require('path');
var pathExists = require('path-exists');
var pull = require('pull-stream');
var toPull = require('stream-to-pull-stream');

function checkArgs(hash, outPath) {
  // format the output directory
  if (!outPath.endsWith('/')) {
    outPath += '/';
  }

  if (!outPath.startsWith('/')) {
    outPath = path.join('/', outPath);
  }

  return outPath;
}

function ensureDir(dir, cb) {
  pathExists(dir).then(function (exists) {
    if (!exists) {
      fs.mkdir(dir, cb);
    } else {
      cb();
    }
  }).catch(cb);
}

function fileHandler(dir) {
  return function onFile(file, cb) {
    var lastSlash = file.path.lastIndexOf('/');
    // Check to see if the result is in a directory
    if (lastSlash === -1) {
      var dirPath = path.join(dir, file.path);
      // Check to see if the result is a directory
      if (file.content) {
        file.content.pipe(fs.createWriteStream(dirPath)).once('error', cb).once('end', cb);
      } else {
        ensureDir(dirPath, cb);
      }
    } else {
      (function () {
        var filePath = file.path.substring(0, lastSlash + 1);
        var dirPath = path.join(dir, filePath);

        ensureDir(dirPath, function (err) {
          if (err) {
            return cb(err);
          }

          if (file.content) {
            var filename = file.path.substring(lastSlash);
            var target = path.join(dirPath, filename);

            file.content.pipe(fs.createWriteStream(target)).once('error', cb).once('end', cb);
            return;
          }
          cb();
        });
      })();
    }
  };
}

module.exports = {
  command: 'get <ipfs-path>',

  describe: 'Download IPFS objects',

  builder: {
    output: {
      alias: 'o',
      type: 'string',
      default: process.cwd()
    }
  },

  handler: function handler(argv) {
    var ipfsPath = argv['ipfs-path'];
    var dir = checkArgs(ipfsPath, argv.output);

    utils.getIPFS(function (err, ipfs) {
      if (err) {
        throw err;
      }

      ipfs.files.get(ipfsPath, function (err, stream) {
        if (err) {
          throw err;
        }
        console.log('Saving file(s) to ' + ipfsPath);
        pull(toPull.source(stream), pull.asyncMap(fileHandler(dir)), pull.onEnd(function (err) {
          if (err) {
            throw err;
          }
        }));
      });
    });
  }
};
