// ipfs-api/lib/api/swarm.js
'use strict';

var promisify = require('promisify-es6');
var multiaddr = require('multiaddr');
var PeerId = require('peer-id');
var PeerInfo = require('peer-info');

module.exports = function (send) {
  return {
    peers: promisify(function (opts, callback) {
      if (typeof opts === 'function') {
        callback = opts;
        opts = {};
      }
      send({
        path: 'swarm/peers',
        qs: opts
      }, function (err, result) {
        if (err) {
          return callback(err);
        }
        callback(null, result.Strings.map(function (addr) {
          return multiaddr(addr);
        }));
      });
    }),
    connect: promisify(function (args, opts, callback) {
      if (typeof opts === 'function') {
        callback = opts;
        opts = {};
      }
      send({
        path: 'swarm/connect',
        args: args,
        qs: opts
      }, callback);
    }),
    disconnect: promisify(function (args, opts, callback) {
      if (typeof opts === 'function') {
        callback = opts;
        opts = {};
      }
      send({
        path: 'swarm/disconnect',
        args: args,
        qs: opts
      }, callback);
    }),
    addrs: promisify(function (opts, callback) {
      if (typeof opts === 'function') {
        callback = opts;
        opts = {};
      }
      send({
        path: 'swarm/addrs',
        qs: opts
      }, function (err, result) {
        if (err) {
          return callback(err);
        }

        var peers = Object.keys(result.Addrs).map(function (id) {
          var info = new PeerInfo(PeerId.createFromB58String(id));
          result.Addrs[id].forEach(function (addr) {
            info.multiaddr.add(multiaddr(addr));
          });

          return info;
        });

        callback(null, peers);
      });
    }),
    localAddrs: promisify(function (opts, callback) {
      if (typeof opts === 'function') {
        callback = opts;
        opts = {};
      }
      send({
        path: 'swarm/addrs/local',
        qs: opts
      }, function (err, result) {
        if (err) {
          return callback(err);
        }
        callback(null, result.Strings.map(function (addr) {
          return multiaddr(addr);
        }));
      });
    })
  };
};
