// ipfs-bitswap/lib/decision/pq.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Heap = require('heap');

module.exports = function () {
  function PriorityQueue(cmp) {
    _classCallCheck(this, PriorityQueue);

    this.q = new Heap(function (a, b) {
      return cmp(a, b) ? -1 : 1;
    });
  }

  _createClass(PriorityQueue, [{
    key: 'push',
    value: function push(e) {
      this.q.push(e);
    }
  }, {
    key: 'pop',
    value: function pop() {
      return this.q.pop();
    }
  }, {
    key: 'update',
    value: function update(e) {
      this.q.updateItem(e);
    }
  }, {
    key: 'size',
    value: function size() {
      return this.q.size();
    }
  }, {
    key: 'isEmpty',
    value: function isEmpty() {
      return this.q.empty();
    }
  }]);

  return PriorityQueue;
}();
